### PR TITLE
defined a more precise type AccessionID

### PR DESCRIPTION
**Pull request by nevrome**
I thought it would be neat to create a data type for accession IDs. That could be the starting point for all sorts of interesting applications, including an automatic download of raw sequencing data.

Looking into this was a confusing experience, though. Reading and validating the ENA IDs is relatively [straight forward](https://ena-docs.readthedocs.io/en/latest/submit/general-guide/accessions.html), but beyond that I'm lost.

1. What are the other relevant archives (DDBJ, NCBI, and...?) and how do their ID systems work? To which degree do the IDs overlap?
2. Which IDs are relevant for us in Poseidon? Should we limit the type of IDs to a more specific subset, e.g. only `Project` and `Sample` IDs?
3. Should we store the project level IDs in the .yml file and only store sample level IDs in the .janno file?

I fear we didn't specify this .janno column sufficiently yet.

### CODE REVIEW

**Comment by nevrome** [comment]
Alright - I just had a beautiful call with @jfy133 to understand these databases a bit better.

I know now, that the neat regexes in the ENA documentation already cover all relevant IDs across the three databases in INSDC (ENA, DDBJ and SRA/NCBI (?)). The database where a sequence was originally submitted is directly encoded in the ID, but it doesn't matter that much, because all these databases mirror each other. My code should therefore be OK -- I just have to replace the prefix `ENA` in the constructors with the more general `INSDC`.

If we ever implement a cool `fetchraw` subcommand to automatically download .bam/.fastq (?) data for Poseidon entries, then we should ask the user to manually select the geographically closest database for downloading.

**Comment by nevrome** [comment]
So 1. is solved and for 3. I realized that our current system of storing this on the individual level is good, because forge pulls the original packages apart. Although having the project-level ID in the POSEIDON.yml file as well might be a useful duplication.

2. is still open.

**Comment by nevrome** [comment]
I experimented with rejecting invalid IDs, but we already have a number of them in the public data repo, so I decided to collect them now as `OtherID`. Shall we be more strict here? Or print a warning?

<details>
  <summary>click here for trident validate on the public repo with the strict rejection of invalid AccessionIDs</summary>

```
[clemens@cranberry published_data]$ trident validate -d .
trident v0.26.4 for poseidon v2.5.0
https://poseidon-framework.github.io

Searching POSEIDON.yml files... 151 found
Checking Poseidon versions... 
Initializing packages... 
> 25 Can't read sample in ./2015_RaghavanScience/2015_RaghavanScience.janno in line 2: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "GSE70987")
Can't read sample in ./2015_RaghavanScience/2015_RaghavanScience.janno in line 3: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "GSE70987")
Can't read sample in ./2015_RaghavanScience/2015_RaghavanScience.janno in line 4: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "GSE70987")
Can't read sample in ./2015_RaghavanScience/2015_RaghavanScience.janno in line 5: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "GSE70987")
Can't read sample in ./2015_RaghavanScience/2015_RaghavanScience.janno in line 6: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "GSE70987")
> 30 Can't read sample in ./2020_Yang_China/Yang_China.janno in line 2: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "10.5281/zenodo.3732838")
Can't read sample in ./2020_Yang_China/Yang_China.janno in line 3: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "10.5281/zenodo.3732838")
Can't read sample in ./2020_Yang_China/Yang_China.janno in line 4: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "10.5281/zenodo.3732838")
Can't read sample in ./2020_Yang_China/Yang_China.janno in line 5: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "10.5281/zenodo.3732838")
Can't read sample in ./2020_Yang_China/Yang_China.janno in line 6: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "10.5281/zenodo.3732838")
> 50 Can't read sample in ./2018_Zalloua_Phoenician/Zalloua_Phoenician.janno in line 2: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "MH43585-43559")
> 65 Can't read sample in ./2016_KanzawaJHG/2016_KanzawaJHG.janno in line 2: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "DRA004001 ")
Can't read sample in ./2016_KanzawaJHG/2016_KanzawaJHG.janno in line 3: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "DRA004001 ")
Can't read sample in ./2016_KanzawaJHG/2016_KanzawaJHG.janno in line 4: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "DRA004001 ")
Can't read sample in ./2016_KanzawaJHG/2016_KanzawaJHG.janno in line 5: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "DRA004001 ")
Can't read sample in ./2016_KanzawaJHG/2016_KanzawaJHG.janno in line 6: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "DRA004001 ")
> 123 Can't read sample in ./2010_RasmussenNature/2010_RasmussenNature.janno in line 2: parse error (Failed reading: conversion error: Not a valid INSDC Accession ID: "SRA010102")
> 151 
Some packages were skipped due to issues:
In the package described in ./2015_RaghavanScience/POSEIDON.yml:
Consistency issues in .janno file ./2015_RaghavanScience/2015_RaghavanScience.janno: Broken lines
In the package described in ./2020_Yang_China/POSEIDON.yml:
Consistency issues in .janno file ./2020_Yang_China/Yang_China.janno: Broken lines
In the package described in ./2018_Zalloua_Phoenician/POSEIDON.yml:
Consistency issues in .janno file ./2018_Zalloua_Phoenician/Zalloua_Phoenician.janno: Broken lines
In the package described in ./2016_KanzawaJHG/POSEIDON.yml:
Consistency issues in .janno file ./2016_KanzawaJHG/2016_KanzawaJHG.janno: Broken lines
In the package described in ./2010_RasmussenNature/POSEIDON.yml:
Consistency issues in .janno file ./2010_RasmussenNature/2010_RasmussenNature.janno: Broken lines
Packages loaded: 146
Validation failed: ERROR
```

</details>

**Comment by jfy133** [comment]
- DRA / SRA prefixes refer to submission IDS - these _should_ thereoretically be associable with the final processed reads
- GSE corresponds to Gene Exprression Omnibus database (normally used more for RNA expression stuff), but still valid - hosted by NCBI (but also I believe EBI under 'ArrayExpress', these should also be reads I think
- The MH* is one type of NCBI Genbank accession prefix but only contains concensus sequences (e.g. mitochondiral genomes). In the Phonecian paper, they made a typo and are missing a leading 0:  MH43585-43559 should be MH043585-043559

**Comment by jfy133** [comment]
Genbank accession suffxies: https://www.ncbi.nlm.nih.gov/genbank/acc_prefix/

File: src/Poseidon/Janno.hs [hunk 1]
```diff
@@ -393,6 +395,63 @@ instance Show RelationDegree where
...
+
+instance Csv.FromField AccessionID where
+    parseField x = do
+        val <- Csv.parseField x
+        pure $ determineAccessionIDType val
+
+-- the patterns are documented at:
+-- https://ena-docs.readthedocs.io/en/latest/submit/general-guide/accessions.html
+determineAccessionIDType :: String -> AccessionID 
+determineAccessionIDType x
```
Comment (stschiff) [inline src/Poseidon/Janno.hs:425]: Very impressive that you've put in all these regexes!

**Review by stschiff — COMMENTED** [review]
OK very nice progress looking at the code. But I think we need to first discuss a bit how we want to specify this at the schema level right? Or shall we simply fix the invalid IDs in the public repo and then roll with this? 

It's also unclear to me what level we want at the sample level. `Sample` would sound correct, but perhaps it's more `analysis` or so, to get the actual sequencing data. Not sure.

**Comment by nevrome** [comment]
Right - the schema is pretty nebulous on the Accession ID column so far. I think we have a case here, where the implementation might help us to figure out, what we actually want to have.

**Comment by stschiff** [comment]
OK, I don't think sufficiently how to actually get data from the ENA... and at which level the BAM- and FASTQ- files (which ultimately is the raw data we want) are stored...

**Comment by stschiff** [comment]
I would like to give some justice to the good work that happened here. From my point of view we can seriously implement this, as long as it doesn't make user experience too painful. I think it's a pretty safe thing because you - ingeniously - have `OtherID` as a constructor as well, so technically this should be a very mild thing.

**Comment by stschiff** [comment]
Coming back to this. I think this is good to merge, too, although we should make sure it doesn't break any workflows with the public repo.

**Pull request merged**